drivers/misc/cldatadrv/rcar_cluster_drv.c

```diff
@@ -0,0 +1,330 @@
+		devm_kfree(&rpdev->dev, event);
+		return -ENOMEM;
+ 	}
+
+	clusterdrv = (rcar_cluster_device_t*)dev_get_drvdata(&rpdev->dev);
+
+	if(!clusterdrv) {
+		dev_err(&rpdev->dev, "send_msg: Can't get data type rcar_cluster_device*\n");
+		devm_kfree(&rpdev->dev, event->result);
+		devm_kfree(&rpdev->dev, event);		
+		return -ENOMEM;
+	}
+
+	msg.Id = cluster_taurus_get_uniq_id();
+	msg.Channel = 0x80;
+	msg.Cmd = R_TAURUS_CMD_IOCTL;
+	msg.Par1 = (uint64_t)data->ioctl_cmd;
+	msg.Par2 = data->value;
+	msg.Par3 = 0;
+
+ 	event->id = msg.Id;
+ 	init_completion(&event->ack);
+	init_completion(&event->completed);
+
+	write_lock(&clusterdrv->event_list_lock);
+	list_add(&event->list, &clusterdrv->taurus_event_list_head);
+	write_unlock(&clusterdrv->event_list_lock);
+
+	ret = rpmsg_send(rpdev->ept, &msg, sizeof(msg));
+	
+	if (ret) {
+		dev_err(&rpdev->dev, "rpmsg_send failed: %d\n", ret);
+		goto end;
+	}
+	do {
+		ret = wait_for_completion_interruptible(&event->ack);
+	}
+	while (ret == -ERESTARTSYS);
+	
+	if(wait_for_completion_interruptible(&event->completed) == -ERESTARTSYS) {
+		dev_err(&rpdev->dev, "send_msg: Interrupted while waiting taurus response (%d)\n", ret);
+		goto end;
+	}
+	else {
+		memcpy(res_msg, event->result, sizeof(taurus_cluster_res_msg_t));
+	}
+
+	write_lock(&clusterdrv->event_list_lock);
+	list_del(&event->list);
+	write_unlock(&clusterdrv->event_list_lock);
+
+end:
+	devm_kfree(&rpdev->dev, event->result);
+	devm_kfree(&rpdev->dev, event);	
+
+	return ret;
+}
+
+/* -----------------------------------------------------------------------------
+ * RPMSG operations
+ */
+
+static int rpmsg_cluster_cb(struct rpmsg_device* rpdev, void* data, int len,
+			void* priv, u32 src) {
+	int ret = 0;
+	struct taurus_event_list* event = NULL;
+	struct taurus_cluster_res_msg* res = (struct taurus_cluster_res_msg*)data;
+	struct list_head* i = NULL;
+	rcar_cluster_device_t* clusterdrv = (rcar_cluster_device_t*)dev_get_drvdata(&rpdev->dev);
+	uint32_t res_id = res->hdr.Id;
+
+	if (!(res->hdr.Result == R_TAURUS_CMD_NOP && res_id == 0)) {/*necessary send data to cluster*/
+		/*send ACK message*/
+		read_lock(&clusterdrv->event_list_lock);
+		
+		list_for_each_prev(i, &clusterdrv->taurus_event_list_head) {
+			event = list_entry(i, struct taurus_event_list, list);
+			if (event->id == res_id) {
+				memcpy(event->result, data, len);
+				if (event->ack_received) {
+					dev_info(&rpdev->dev, "send_msg: Message completed (%d)\n", ret);
+					complete(&event->completed);
+				} else {
+					event->ack_received = 1;
+					complete(&event->ack);
+				}
+			}
+		}
+		read_unlock(&clusterdrv->event_list_lock);
+	}
+	return 0;
+}
+
+static int rpmsg_cluster_probe(struct rpmsg_device* rpdev)
+{
+	rcar_cluster_device_t *clusterdvc = NULL;
+	int ret = 0;
+	struct miscdevice *dev = NULL;
+	taurus_cluster_res_msg_t res_msg;
+
+	taurus_cluster_data_t cluster_data = {
+		.value = 10,
+		.ioctl_cmd = 1,
```

Comment:
What those values mean?

Reply:
just ping to test communication, can be removed